flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/AbstractAsyncStateStreamOperatorV2.java

```diff
@@ -259,17 +265,35 @@ public void processWatermark(Watermark mark) throws Exception {
             super.processWatermark(mark);
             return;
         }
-        asyncExecutionController.processNonRecord(
-                () -> {
-                    // todo: make async operator deal with interruptible watermark
-                    if (timeServiceManager != null) {
-                        CompletableFuture<Void> future = timeServiceManager.advanceWatermark(mark);
-                        future.thenAccept(v -> output.emitWatermark(mark));
-                        asyncExecutionController.drainWithTimerIfNeeded(future);
-                    } else {
+        asyncExecutionController.processNonRecord(() -> doProcessWatermark(mark, null));
+    }
+
+    /**
+     * Handle the watermark and timers, then run a provided {@link Runnable} asynchronously right
+     * after the watermark is emitted.
+     *
+     * @param mark The watermark.
+     * @param postAction The runnable for post action.
+     */
+    protected void doProcessWatermark(Watermark mark, @Nullable Runnable postAction)
+            throws Exception {
+        // todo: make async operator deal with interruptible watermark
+        if (timeServiceManager != null) {
+            CompletableFuture<Void> future = timeServiceManager.advanceWatermark(mark);
+            future.thenAccept(
+                    v -> {
                         output.emitWatermark(mark);
-                    }
-                });
+                        if (postAction != null) {
+                            postAction.run();
+                        }
+                    });
+            asyncExecutionController.drainWithTimerIfNeeded(future);
+        } else {
+            output.emitWatermark(mark);
```

Comment:
nit: the future.thenAccept( and the else seem to have the same logic, can we put this in a function?

Reply:
Sure, done